From: http://xenbits.xensource.com/linux-2.6.18-xen.hg
# HG changeset 540 patch
# User Keir Fraser <keir.fraser@citrix.com>
# Date 1210670928 -3600
# Node ID 3044873a84b70e7bbae977037ef97fe18670e166
# Parent  29b8c3f366031a6f047777da6be0bed9b307ad5a
Subject: Avoid theoretical TOCTTOU bug in block backend nr_segments checking.

Based on a patch by Steven Smith <steven.smith@citrix.com>

Signed-off-by: Keir Fraser <keir.fraser@citrix.com>
Acked-by: jbeulich@novell.com

Index: 11.0-2008-06-05/drivers/xen/blkback/blkback.c
===================================================================
--- 11.0-2008-06-05.orig/drivers/xen/blkback/blkback.c	2008-06-05 11:27:46.000000000 +0200
+++ 11.0-2008-06-05/drivers/xen/blkback/blkback.c	2008-06-05 11:29:05.000000000 +0200
@@ -344,6 +344,9 @@ static int do_block_io_op(blkif_t *blkif
 		}
 		blk_rings->common.req_cons = ++rc; /* before make_response() */
 
+		/* Apply all sanity checks to /private copy/ of request. */
+		barrier();
+
 		switch (req.operation) {
 		case BLKIF_OP_READ:
 			blkif->st_rd_req++;
Index: 11.0-2008-06-05/drivers/xen/blktap/blktap.c
===================================================================
--- 11.0-2008-06-05.orig/drivers/xen/blktap/blktap.c	2008-06-05 11:27:46.000000000 +0200
+++ 11.0-2008-06-05/drivers/xen/blktap/blktap.c	2008-06-05 11:29:05.000000000 +0200
@@ -1264,6 +1264,9 @@ static int do_block_io_op(blkif_t *blkif
 		}
 		blk_rings->common.req_cons = ++rc; /* before make_response() */
 
+		/* Apply all sanity checks to /private copy/ of request. */
+		barrier();
+
 		switch (req.operation) {
 		case BLKIF_OP_READ:
 			blkif->st_rd_req++;
Index: 11.0-2008-06-05/include/xen/blkif.h
===================================================================
--- 11.0-2008-06-05.orig/include/xen/blkif.h	2008-06-05 10:59:39.000000000 +0200
+++ 11.0-2008-06-05/include/xen/blkif.h	2008-06-05 11:29:05.000000000 +0200
@@ -98,8 +98,9 @@ static void inline blkif_get_x86_32_req(
 	dst->handle = src->handle;
 	dst->id = src->id;
 	dst->sector_number = src->sector_number;
-	if (n > src->nr_segments)
-		n = src->nr_segments;
+	barrier();
+	if (n > dst->nr_segments)
+		n = dst->nr_segments;
 	for (i = 0; i < n; i++)
 		dst->seg[i] = src->seg[i];
 }
@@ -112,8 +113,9 @@ static void inline blkif_get_x86_64_req(
 	dst->handle = src->handle;
 	dst->id = src->id;
 	dst->sector_number = src->sector_number;
-	if (n > src->nr_segments)
-		n = src->nr_segments;
+	barrier();
+	if (n > dst->nr_segments)
+		n = dst->nr_segments;
 	for (i = 0; i < n; i++)
 		dst->seg[i] = src->seg[i];
 }
